### PR TITLE
ETL ADR

### PR DESCRIPTION
**Pull request by pomadchin**
- [ ] another version with a single json file
- [ ] split adrs into sep files
- [ ] fix urls

ADR v2 (just not to create tons of PR's and to keep history): https://github.com/pomadchin/geotrellis/blob/adr/etl-v2/docs/architecture/004-etl-pipeline.rst

Related to #1767

### CODE REVIEW

File: docs/architecture/adr.rst [hunk 1]
```diff
@@ -439,6 +439,190 @@ Conclusion
...
 that can lead to significant performance loss.
+
+0004 - ETL Pipelne
+------------------
+
+Context
+^^^^^^^
+
+Current GeoTrellis ETL does not allow to determine ETL as a pipeline of transformation / actions.
+This document would describe a new approach(inspired by [PDAL Pipeline](https://www.pdal.io/pipeline.html))
```
Comment (fosskers) [inline docs/architecture/adr.rst]: `would describe` -> `describes`

File: docs/architecture/adr.rst [hunk 2]
```diff
@@ -439,6 +439,190 @@ Conclusion
...
+used threads more than CPU cores available for java machine, otherwise
 that can lead to significant performance loss.
+
+0004 - ETL Pipelne
+------------------
+
+Context
+^^^^^^^
+
+Current GeoTrellis ETL does not allow to determine ETL as a pipeline of transformation / actions.
```
Comment (fosskers) [inline docs/architecture/adr.rst]: `Current` -> `The current`

`allow to determine` -> `allow us to define`

File: docs/architecture/adr.rst [hunk 3]
```diff
@@ -439,6 +439,190 @@ Conclusion
...
+This document would describe a new approach(inspired by [PDAL Pipeline](https://www.pdal.io/pipeline.html))
+with a new ETL JSON description.
+
+Decision
+^^^^^^^^
+
+We can divide current ETL into following steps:
```
Comment (fosskers) [inline docs/architecture/adr.rst]: `divide current` -> `divide the current`

`into following` -> `into the following`

File: docs/architecture/adr.rst [hunk 4]
```diff
@@ -439,6 +439,190 @@ Conclusion
...
+
+* Load
+* Reproject
+* Tile
+* Pyramid
+* Render
+* Save
+
+It is possible to represent all these steps as JSON objects, and an array of these objects would be an ETL instructions pipeline.
```
Comment (fosskers) [inline docs/architecture/adr.rst]: `ELT instructions` -> `ETL instruction`

File: docs/architecture/adr.rst [hunk 5]
```diff
@@ -439,6 +439,190 @@ Conclusion
...
+There still would be three different json inputs: ``input.json``, ``output.json`` and ``backend-profiles.json`` as it seems to be
+a reasonable way to divide parameters semantically.
+
+
+input.json
+^^^^^^^^^^
+
+The input schema, would be mostly without changes. The only difference would be in the field ``name``,
```
Comment (fosskers) [inline docs/architecture/adr.rst]: `schema, would` -> `schema would`

File: docs/architecture/adr.rst [hunk 6]
```diff
@@ -439,6 +439,190 @@ Conclusion
...
+which should be added as a CLI argument, as an override option. That would allow to skip some unnecessary
```
Comment (fosskers) [inline docs/architecture/adr.rst]: `allow to` -> `allow us to`

File: docs/architecture/adr.rst [hunk 7]
```diff
@@ -439,6 +439,190 @@ Conclusion
...
+machinery, in case of all inputs should have the same name (handled as a single layer). Another option,
```
Comment (fosskers) [inline docs/architecture/adr.rst]: `in case of all inputs should have` -> `in the case where all inputs have`

File: docs/architecture/adr.rst [hunk 8]
```diff
@@ -439,6 +439,190 @@ Conclusion
...
+is to have ``name`` field as an optional field of a ``{save | update | reindex}`` pipeline step.
```
Comment (fosskers) [inline docs/architecture/adr.rst]: `to have name field` -> `to have the name field`

File: docs/architecture/adr.rst [hunk 9]
```diff
@@ -439,6 +439,190 @@ Conclusion
...
+
+
+backend-profiles.json
+^^^^^^^^^^^^^^^^^^^^^
+
+Without significant changes, as it is already provide minimal information about backends credentials.
```
Comment (fosskers) [inline docs/architecture/adr.rst]: `as it is already provide` -> `as it already provides`

File: docs/architecture/adr.rst [hunk 10]
```diff
@@ -439,6 +439,190 @@ Conclusion
...
+        "path": "path or table",
+        "profile": "profile name"
+      }
+    }
+  ]
+
+Conclusion
+^^^^^^^^^^
+
+Current ``input.json`` and ``backend-profiles.json`` are seems to be already fine. Significant changes should be introduced
```
Comment (fosskers) [inline docs/architecture/adr.rst]: `Current` -> `The current`

`are seems to be already fine` -> `seem to be fine as-is`

File: docs/architecture/adr.rst [hunk 11]
```diff
@@ -439,6 +439,190 @@ Conclusion
...
+into ``output.json`` as specified above. That would allow to construct Pipelines similar to what PDAL allows. In addition,
```
Comment (fosskers) [inline docs/architecture/adr.rst]: `allow to` -> `allow us to`

File: docs/architecture/adr.rst [hunk 12]
```diff
@@ -439,6 +439,190 @@ Conclusion
...
+such approach allows to have not complicated API extensions, which can be implemented just by implementing desired
```
Comment (fosskers) [inline docs/architecture/adr.rst]: `such approach...` -> `such an approach allows us to not have complicated...`

**Comment by pomadchin** [comment]
Thank you Colin!

**Comment by echeipesh** [comment]
Superseded by: https://github.com/locationtech/geotrellis/pull/2264

**Pull request closed without merging**